Lib/fontTools/ttLib/tables/_k_e_r_n.py

```diff
@@ -153,6 +158,7 @@ def decompile(self, data, ttFont):
 	def compile(self, ttFont):
 		nPairs = len(self.kernTable)
 		searchRange, entrySelector, rangeShift = getSearchRange(nPairs, 6)
+		searchRange = min(searchRange, 32768)
```

Comment:
@moyogo why did you clamp searchRange to maximum 32768?

searchRange is defined as:

> The largest power of two less than or equal to the value of nPairs, multiplied by the size in bytes of an entry in the table.

The size of an entry is 6. The searchRange is a uint16, so it cannot exceed 0xFFFF (65535).
I think the maximum searchRange that we should allow is `(2 ** 13) * 6`, which is 0xC000 (49152).
The next power of two would be bigger than the maximum unsigned short: `(2 ** 14) * 6 = 0x18000` (98304).

Reply:
actually, it doesn't make sense to only clamp searchRange, as the latter always goes hand in hand with the other two (entrySelector and rangeShift).

The maximum number of pairs that still produce a valid searchRnge, entrySelector and rangeShift triplet is 0x3FFF (16383). Anything greater than that will overflow USHORT

```python
>>> from fontTools.ttLib import getSearchRange
>>> getSearchRange(16383, 6)
(49152, 13, 49146)

>>> getSearchRange(16384, 6)
(98304, 14, 0)


```

Reply:
Indeed, I should have looked at this better we don’t want to clamp at 32768. OpenSans and calibri fit 0x18000 into 0xFFFF, which is 0x8000. So I’ll do searchRange = searchRange % 0x10000.

Reply:
if you use modulo like that, then you may be setting searchRange to an incorrect value. 65535 cannot be divided by 6 (the size of the kern pair records).
I am not sure whether those three related fields, searchRange, entrySelector and rangeShift are there only for optimizing the binary search, or what would happen if they don't match exactly the wording of the spec. That's why I suggested to not allow more than 16383, which is the maximum number that still produces a valid searchRnge, entrySelector and rangeShift triplet (one that does not overflow USHORT)

Reply:
>> I’ll do searchRange = searchRange % 0x10000.

> if you use modulo like that, then you may be setting searchRange to an incorrect value. 

@moyogo sorry, I need to correct myself. Doing `n % 0x10000` as you originally suggested does exactly the same thing as doing `n & 0xFFFF`(i.e. truncates n to a 16-bit unsigned integer). Apologies for the confusion.